File: 079_word_search.rb
=begin
Given a 2D board and a word, find if the word exists in the grid.

The word can be constructed from letters of sequentially adjacent cell, where "adjacent" cells are those horizontally or vertically neighboring. The same letter cell may not be used more than once.

Example:

board =
[
  ['A','B','C','E'],
  ['S','F','C','S'],
  ['A','D','E','E']
]

Given word = "ABCCED", return true.
Given word = "SEE", return true.
Given word = "ABCB", return false.
=end

# @param {Character[][]} board
# @param {String} word
# @return {Boolean}
def exist(board, word)
    board.each_with_index do |row, i|
        row.each_with_index do |char, j|
            visited = []
            if search(word, board, i, j)
                return true
            end
        end
    end
    
    return false
end

def search(word, board, i, j)
    if word.length == 0
        return true
    end

    if out_of_bounds(board, i, j) || word[0] != board[i][j]
        return false
    end

    tmp = board[i][j]
    board[i][j] = nil
    word = word[1..-1]

    exists = search(word, board, i+1, j) ||
        search(word, board, i-1, j) ||
        search(word, board, i, j+1) ||
        search(word, board, i, j-1)
    board[i][j] = tmp
    
    return exists
end

def out_of_bounds(board, i, j)
   return i < 0 || j < 0 || i >= board.length || j >= board[0].length 
end

File: 079_word_search.go
/*
Given an m x n grid of characters board and a string word, return true if word exists in the grid.

The word can be constructed from letters of sequentially adjacent cells, where adjacent cells are horizontally or vertically neighboring. The same letter cell may not be used more than once.

 

Example 1:


Input: board = [["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]], word = "ABCCED"
Output: true
Example 2:


Input: board = [["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]], word = "SEE"
Output: true
Example 3:


Input: board = [["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]], word = "ABCB"
Output: false
 

Constraints:

m == board.length
n = board[i].length
1 <= m, n <= 6
1 <= word.length <= 15
board and word consists of only lowercase and uppercase English letters.
 

Follow up: Could you use search pruning to make your solution faster with a larger board?
*/

func exist(board [][]byte, word string) bool {
    for i := range board {
        for j := range board[0] {
            if dfs(board, word, i, j) {
                return true
            }
        }
    }
    return false
}

func dfs(board [][]byte, word string, i, j int) bool {
    if len(word) == 0 {
        return true
    }
    if i < 0 || i >= len(board) || j < 0 || j >= len(board[0]) {
        return false
    }
    if word[0] != board[i][j] {
        return false
    }
    l := board[i][j]
    board[i][j] = '@'
    if dfs(board, word[1:], i-1, j) ||
        dfs(board, word[1:], i+1, j) ||
        dfs(board, word[1:], i, j-1) ||
        dfs(board, word[1:], i, j+1) {
        return true
    }
    board[i][j] = l
    return false
}

